src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSLambdaHttpUtils.cs

```diff
@@ -30,40 +33,70 @@ internal static IEnumerable<KeyValuePair<string, object>> GetHttpTags<TInput>(TI
-                var path = request.RequestContext?.Path ?? request.Path ?? string.Empty;
-                httpTarget = string.Concat(path, GetQueryString(request));
+                urlPath = request.RequestContext?.Path ?? request.Path ?? string.Empty;
+                urlQuery = GetQueryString(request);
+                httpTarget = string.Concat(urlPath, urlQuery);
                 httpMethod = request.RequestContext?.HttpMethod ?? request.HttpMethod;
                 var hostHeader = AWSLambdaUtils.GetHeaderValues(request, HeaderHost)?.LastOrDefault();
                 (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeader);
                 break;
             case APIGatewayHttpApiV2ProxyRequest requestV2:
                 httpScheme = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderXForwardedProto)?.LastOrDefault();
-                httpTarget = string.Concat(requestV2.RawPath ?? string.Empty, GetQueryString(requestV2));
+                urlPath = requestV2.RawPath ?? string.Empty;
+                urlQuery = GetQueryString(requestV2);
+                httpTarget = string.Concat(urlPath, urlQuery);
                 httpMethod = requestV2.RequestContext?.Http?.Method;
                 var hostHeaderV2 = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderHost)?.LastOrDefault();
                 (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeaderV2);
                 break;
             case ApplicationLoadBalancerRequest albRequest:
                 httpScheme = AWSLambdaUtils.GetHeaderValues(albRequest, HeaderXForwardedProto)?.LastOrDefault();
-                httpTarget = string.Concat(albRequest.Path ?? string.Empty, GetQueryString(albRequest));
+                urlPath = albRequest.Path ?? string.Empty;
+                urlQuery = GetQueryString(albRequest);
+                httpTarget = string.Concat(urlPath, urlQuery);
                 httpMethod = albRequest.HttpMethod;
                 var albHostHeader = AWSLambdaUtils.GetHeaderValues(albRequest, HeaderHost)?.LastOrDefault();
                 (hostName, hostPort) = GetHostAndPort(httpScheme, albHostHeader);
                 break;
             default:
-                return tags;
+                return Enumerable.Empty<KeyValuePair<string, object>>();
+        }
+
+        var tags = semanticConventions.AttributeBuilder;
+
+        if (httpScheme != null)
+        {
+            tags.AddAttributeHttpScheme(httpScheme, addIfEmpty: true);
+            tags.AddAttributeUrlScheme(httpScheme, addIfEmpty: true);
+        }
+
+        tags.AddAttributeHttpTarget(httpTarget, addIfEmpty: true);
+        tags.AddAttributeUrlPath(urlPath, addIfEmpty: true);
+
+        if (urlQuery != null)
+        {
+            tags.AddAttributeUrlQuery(urlQuery, addIfEmpty: true);
+        }
+
+        if (httpMethod != null)
+        {
+            tags.AddAttributeHttpMethod(httpMethod, addIfEmpty: true);
```

Comment:
method vs request method is a matter of semantic convention version, so I would expect to have a single method (probably named after the newer semantic convention version) that does the right thing(s) depending on version.

There may be cases where the values differ depending on semconv version and then it makes sense to require different calls per version (and make some calls do nothing depending on semconv version) but does this apply to http method vs http request method?

Reply:
> There may be cases where the values differ depending on semconv version and then it makes sense to require different calls per version

You ask a good question.  I've been working to balance the maintainability of the `AWSSemanticConventions` class, so that's it easy to track how Semantic Convention changes over time and that the AddAttributeXXX methods continue to align with the actual attribute name; vs keeping the business logic class, like `AWSLambdaHttpUtils` free from needing to deal with supporting multiple Semantic Convention versions.

I originally implemented Http Method as you suggested, where `AWSSemanticConventions_v1_29_0` changed the name of `AttributeHttpMethod` to `"http.request.method"`.

However, **I decided to change course for any Attributes that are clearly marked Obsolete.**

As [`HttpAttributes.AttributeHttpMethod`](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.SemanticConventions/Attributes/HttpAttributes.cs#L43-L44) is marked Obsolete in the spec, trying to adapt the `AddAttributeHttpMethod()` felt like it would cause more maintainability problems down the road vs marking that method obsolete and updating `AWSLambdaHttpUtils` to make an additional call to `AddAttributeHttpRequestMethod`.

A technical note: `AWSSemanticConventions.Add` does check if the `attributeName` is empty and does not add it.  So v1.27 will only output `"http.method"` and v1.29+ will only output `"http.request.method"`.  They're never both output.

Thoughts?

Reply:
I don't think it's ideal, but probably good enough. Instrumentation developers need to know all the different versions they support and make sure to set all the relevant aliases/variants, otherwise the function will silently do nothing. For something only used locally within the AWS instrumentations I think it's acceptable, if we want to use that more widely, it would IMHO need reconsideration.

Reply:
@Oberon00 - I agree with both of your points.  For this PR, I think we should move forward.  Fortunately, everything is `internal` so we can tweak this implementation moving forward to best support maintainability.

Reply:
@alanwest - 

Yes, the very first entry in `SemanticConventionsVersions` is a little odd.  I intended it to represent the current Semantic Conventions being used, which is really a mixture of many different versions.

Originally, `SemanticConventionsVersions` tracked the version of the OpenTelemetry.*.AWS libraries themselves, rather then a specific Semantic Convention, so this aspect made a bit more sense.  But after some [discussion](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2367#discussion_r1873338621) we decided to track the Semantic Convention version instead.

**I'm very open to suggestions on how to tweak naming / documentation to make this clearer. **

I could also upgrade the library to use the full version of 1.27 - but I don' think that would help any users:  anyone who wants the latest convention should use `1.29` and anyone that wants backwards compatibility with the existing betas wouldn't have an option.

Perhaps we could rename `V1_27_0_Experimental` -> `Legacy_Experimental`.  But this feels like it might be more confusing then it's worth.

**Thoughts?**

Reply:
> Perhaps we could rename V1_27_0_Experimental -> Legacy_Experimental. But this feels like it might be more confusing then it's worth.

Yes, I was thinking the same kind of thing.. one idea is that you could release things as follows

**Next non-stable release**

```csharp
enum SemanticConventionVersions
{
    Legacy = -1, // This would reflect the telemetry as it is today and be the default while not stable
    Latest,
    V1_29_0_Experimental,
}
```

** Future release candidate / stable releases **

Drop the `Legacy` option. This is a breaking change, but that's ok. Folks who still depend on the legacy behavior can continue to use prior non-stable versions until they are ready to upgrade.

```csharp
enum SemanticConventionVersions
{
    Latest,
    V1_29_0_Experimental, // This would become the new default
}
```

We had a release plan very similar to this when we stabilized ASP.NET Core and HttpClient instrumentation. The last non-stable release defaulted to the legacy conventions, though you could opt in to the new HTTP conventions using the `OTEL_SEMCONV_STABILITY_OPT_IN` environment variable. Then, when we shipped the instrumentation stable, we removed the support for `OTEL_SEMCONV_STABILITY_OPT_IN` and defaulted to the new, stable HTTP conventions.

Reply:
@alanwest - I've made a number of changes based on your feedback.  It deviates slightly but I think captures the spirit of your suggestions:

### `Legacy`
I've dropped support for `Legacy` Semantic Conventions.  This made more sense when the design was for `SemanticConventionVersions` to track the version of the OpenTelemetry.*.AWS Libraries.  But now that I've switched to tracking the OpenTelemetry.SemanticConventions instead, `Legacy` is more complicated then it's worth.  

This addresses your concern of `1.27` not _really_ being `1.27`.  

And, as you pointed out, "Folks who still depend on the legacy behavior can continue to use prior non-stable versions until they are ready to upgrade."

### Experimental
During the SIG meeting discussion, I walked away that there was a strong desire amongst the quorum to communicate to users that this GA library was using Experimental conventions.  My opinion is it will be difficult to communicate to users that some Semantic Convention is stable, while others are experimental; it's safer to keep `Experimental` in the name up until all Conventions used by the OpenTelemetry.*.AWS Libraries are stable.

### Release Plan

Dropping `Legacy` simplifies things:

**_1.10.0-beta.next_**

```
enum SemanticConventionVersions
{
    Latest,
    V1_28_0_Expiremental,  // Default
    V1_29_0_Expiremental
}
```

_**1.10.0** // first prod release_

```
enum SemanticConventionVersions
{
    Latest,
    V1_28_0_Expiremental,  // Default
    V1_29_0_Expiremental
}
```

_**1.11.0** // vNext beta/prod_
```
enum SemanticConventionVersions
{
    Latest,
    V1_28_0_Expiremental,  // Default
    V1_29_0_Expiremental,
    v1_32_0_Expiormental // assuming 1.32 is the most recent Sem Conv come release time for 1.11
}
```

The library continues to default to `V1_28_0_Expiremental` to offer users stability, but continues to add support for (some) newer Semantic Conventions for users that wish to opt-in.

**Is this something you can support?**

Reply:
> During the SIG meeting discussion, I walked away that there was a strong desire amongst the quorum to communicate to users that this GA library was using Experimental conventions.

Communication is important, though I do not think adding `Experimental` to the names of the enum values help. If anything, it makes things more confusing. I think you can do everything you need to communication wise through documentation. I like the pattern this PR is proposing. If we were to apply it to other instrumentation in the contrib repo that I am a code owner for, I will strongly advocate not suffixing the name of versions with `Experimental`. I think the versions speak clearly enough for themselves.

In the SIG meeting, I emphasized the thing of primary importance to end users. It essentially boils down to: If you declare instrumentation stable then the telemetry that instrumentation produces must not change by default in a breaking way until a major version bump. From the standpoint of my support and review of this PR, this is the only lens I'm reviewing it through. I am not familiar with the AWS instrumentation to weigh in on any other aspect, so I'm relying on the approval and support of the code owners of these components. Though, personally, I support your approach. I think it meets the requirements of the specification with regards to telemetry stability.

As for your release plan, it's up to you which versions of the conventions you want to initially support. However, I'm still not sure I understand why you have chosen to support 1.28 and 1.29 to begin with. It seems to me the simplest thing would be to initially support only whatever the latest version is at the time of your stable release.

---

As I've thought about this PR a bit more, I want to call out one more additional consideration that came to mind. Here's a potential limitation of the approach you're taking:

Let's say you ship stable and default to the v1.28 of the conventions. Now imagine two things happen when v1.32 of the conventions are released:

1. A new attribute is introduced for HTTP spans. Adding span attributes is considered a non-breaking change, so it is allowed for instrumentation to implement this change without a major version bump.
2. The faas conventions evolve in breaking ways and are still not yet considered stable.

Let's say an end user wants the new HTTP attribute but is not yet ready for the breaking changes to the faas telemetry. They're essentially stuck because selecting `V1_32_0` 

I personally do not think this limitation is a showstopper, but I just wanted to call it out to make sure you agree this is a consequence of the approach you're taking and to bring more clarity to others reviewing this PR.

Reply:
**"Experimental"**
I think your argument about the Enum not including "Experimental" make sense.  Upon reflection, I think it best to defer to your experienced opinion in this matter.  I've updated the Enum to drop the "Experimental" suffix.

**Semantic Convention Granularity**
Per the call out of the SemanticConventionVersions not being fine grained enough to allow selecting the version of each Semantic Convention document, I did think through this when designing.  Because the AWS libraries use ~10 different documents, I felt offering fine grained control, while technically feasible, would be too overwhelming for users and it would be a better approach to stick with a single value in the Options class.  

If there was sufficient customer demand for better control, we could offer it in a future version.  

But in the end, hopefully this is all just a short term measure until Semantic Convention stabilizes.